scalapact-core/src/main/scala/com/itv/scalapactcore/common/matching/InteractionMatchers.scala

```diff
@@ -13,52 +13,34 @@ import scala.xml._
 
 object InteractionMatchers {
 
-  import StatusMatching._
-  import PathMatching._
-  import HeaderMatching._
-  import MethodMatching._
-  import BodyMatching._
-
   lazy val matchRequest: Boolean => List[Interaction] => InteractionRequest => Either[String, Interaction] = strictMatching => interactions => received =>
-    interactions.find { ir =>
-      if(strictMatching) isStrictRequestMatch(ir)(received)
-      else isNonStrictRequestMatch(ir)(received)
+    interactions.find { interaction =>
+      InteractionMatchingPrograms
+        .matchRequestProgram(interaction.request, received)
+        .foldMap {
+          if(strictMatching) MatchingInterpreters.Request.strict
+          else MatchingInterpreters.Request.permissive
```

Comment:
This is kind of the important bit. What happens here is that we take a general **process** but apply a different **strategy** for it's evaluation. I've come to think of the Free Monad (pattern?) as being a lot like the good old strategy patterns of OO, expect that the process, the steps needed to achieve the result don't change, only what you might do at each step.